components/ws-daemon/pkg/controller/workspace_operations.go

```diff
@@ -0,0 +1,492 @@
+	// currently we're only uploading prebuild log files
+	logFiles, err := logs.ListPrebuildLogFiles(ctx, opts.WorkspaceLocation)
+	if err != nil {
+		return err
+	}
+
+	rs, err := storage.NewDirectAccess(&wso.config.Storage)
+	if err != nil {
+		return xerrors.Errorf("cannot use configured storage: %w", err)
+	}
+
+	err = rs.Init(ctx, opts.Meta.Owner, opts.Meta.WorkspaceId, opts.Meta.InstanceId)
+	if err != nil {
+		return xerrors.Errorf("cannot use configured storage: %w", err)
+	}
+
+	err = rs.EnsureExists(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, absLogPath := range logFiles {
+		taskID, parseErr := logs.ParseTaskIDFromPrebuildLogFilePath(absLogPath)
+		if parseErr != nil {
+			glog.WithError(parseErr).Warn("cannot parse headless workspace log file name")
+			continue
+		}
+
+		err = retryIfErr(ctx, 5, glog.WithField("op", "upload log"), func(ctx context.Context) (err error) {
+			_, _, err = rs.UploadInstance(ctx, absLogPath, logs.UploadedHeadlessLogPath(taskID))
+			if err != nil {
+				return
+			}
+
+			return
+		})
+		if err != nil {
+			return xerrors.Errorf("cannot upload workspace logs: %w", err)
+		}
+	}
+	return err
+}
+
+func (wso *WorkspaceOperations) uploadWorkspaceContent(ctx context.Context, sess *session.Workspace) error {
+	var backupName = storage.DefaultBackup
+	// Avoid too many simultaneous backups in order to avoid excessive memory utilization.
+	waitStart := time.Now()
+	select {
+	case wso.backupWorkspaceLimiter <- struct{}{}:
+	case <-time.After(15 * time.Minute):
+		// we timed out on the rate limit - let's upload anyways, because we don't want to actually block
+		// an upload. If we reach this point, chances are other things are broken. No upload should ever
+		// take this long.
+		wso.metrics.BackupWaitingTimeoutCounter.Inc()
+	}
+
+	waitTime := time.Since(waitStart)
+	wso.metrics.BackupWaitingTimeHist.Observe(waitTime.Seconds())
+
+	defer func() {
+		// timeout -> we did not add to the limiter
+		if waitTime.Minutes() >= 15 {
+			return
+		}
+
+		<-wso.backupWorkspaceLimiter
+	}()
+
+	var (
+		loc  = sess.Location
+		opts []storage.UploadOption
+	)
+
+	err := os.Remove(filepath.Join(sess.Location, wsinit.WorkspaceReadyFile))
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		// We'll still upload the backup, well aware that the UX during restart will be broken.
+		// But it's better to have a backup with all files (albeit one too many), than having no backup at all.
+		glog.WithError(err).WithFields(sess.OWI()).Warn("cannot remove workspace ready file")
+	}
+
+	rs, ok := sess.NonPersistentAttrs[session.AttrRemoteStorage].(storage.DirectAccess)
+	if rs == nil || !ok {
+		return xerrors.Errorf("no remote storage configured")
+	}
+
+	var (
+		tmpf     *os.File
+		tmpfSize int64
+	)
+
+	defer func() {
+		if tmpf != nil {
+			// always remove the archive file to not fill up the node needlessly
+			os.Remove(tmpf.Name())
+		}
+	}()
```

Comment:
thought I left a comment but got lost, but this should probably also be done inside the retry func if there was an error, otherwise only the last retry will get cleaned up